### PR TITLE
Resolved mismatch between DB and file system caused by download cancellation

### LINKED ISSUE #1692: FileNotFoundException after cancelling and re-downloading a large form
#### Software and hardware versions 
Collect v1.11.0

#### Problem description
FileNotFoundException after cancelling and re-downloading a large form (that is already present on the device).

#### Steps to reproduce the problem
Download a large form successfully. Start, and then cancel a second download. Download it again right away. Wait for FileNotFoundException.

#### Other information 
After discussion here: https://opendatakit.slack.com/archives/C35ENHA64/p1517262884000576, it seems it is best to not start a new DownloadFormsTask (AsyncTask) until the previous one completes, since they are accessing the same forms, file system and database. Even after canceling a download, the task is blocked here: 
`parsedFields = FileUtils.parseXML(fileResult.file);`

Perhaps the progress dialog should change to `Canceling...` or something and continue to block until the download task is finished. We could also look at how to avoid the parsing blocking for a long time, perhaps by doing it in another thread, and abandoning it or killing it on cancellation.

### PR DESCRIPTION
**Pull request by lakshyagupta21**
Closes #1692 

#### What has been done to verify that this works as intended?
Tried on Physical Device (Android 5.0 and 6.0) and behaviour found correct

#### Why is this the best possible solution? Were any other approaches considered?
Row from the database should be removed whenever file is deleted from the filesystem, whenever cleanUp() is called first it removes from database and then from the filesystem.

#### Are there any risks to merging this code? If so, what are they?
No
#### Do we need any specific form for testing your changes? If so, please attach one.
I used AllWidgets form for the testing.

### CODE REVIEW

**Comment by dcbriccetti** [comment]
Hi. Thanks for this. Can you explain what was happening, and why one had to start the second download immediately to cause the program, rather than waiting a while after cancelling.

**Comment by dcbriccetti** [comment]
I don’t think this change worked.
[failure.txt](https://github.com/opendatakit/collect/files/1675279/failure.txt)

**Comment by lakshyagupta21** [comment]
Actually this code was deleting the newly downloaded file because file row already exists in database so I changed cleanUp code method so that it deletes the entry from the database also.

```
try {
            c = formsDao.getFormsCursorForMd5Hash(FileUtils.getMd5Hash(f));
            if (c.getCount() > 0) {
                // Should be at most, 1
                c.moveToFirst();

                isNew = false;

                // delete the file we just downloaded, because it's a duplicate
                Timber.w("A duplicate file has been found, we need to remove the downloaded file "
                        + "and return the other one.");
                FileUtils.deleteAndReport(f);

                // set the file returned to the file we already had
                String existingPath = c.getString(c.getColumnIndex(FormsColumns.FORM_FILE_PATH));
                f = new File(existingPath);
                Timber.w("Will use %s", existingPath);
            }
```

**Comment by lakshyagupta21** [comment]
It would be helpful for debugging if you share the form that you are testing on.

**Review by dcbriccetti — CHANGES_REQUESTED** [review]
We’ve made a new plan, so I expect @lakshyagupta21 will be replacing this PR with a new one.

**Comment by lakshyagupta21** [comment]
@dcbriccetti Please have a look at the code I've updated the PR.

File: collect_app/src/main/res/values/strings.xml
Comment (dcbriccetti) [inline collect_app/src/main/res/values/strings.xml]: We should consistently use one or the other. Do we usually use U.S. spellings? Canceling, if so?

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java]: What do you think about moving this code elsewhere? This module is very long already. Maybe formsDao.deleteFormsFromMd5Hash? That would also make it clear at a glance what these 12 lines do.

File: collect_app/src/main/res/values/strings.xml
Comment (lakshyagupta21) [reply collect_app/src/main/res/values/strings.xml]: Its is a typo error, I think we use U.S. Spellings.

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java
Comment (lakshyagupta21) [reply collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java]: That would be good idea to have one proper method to remove file row from DB on the basis of its Md5Hash value.

**Comment by dcbriccetti** [comment]
That’s a nice improvement. I have some more comments coming.

File: collect_app/src/main/java/org/odk/collect/android/dao/FormsDao.java
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/dao/FormsDao.java]: Using varargs will simplify calling the method. Use new style for. Since you have a loop, you need to close the cursor each time.

delete forms changes

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java]: You can now simplify to
`formsDao.deleteFormsFromMd5Hash(FileUtils.getMd5Hash(fileResult.file));`

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: You might consider having the title be “Canceling” and have the message be please_wait.

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: The purpose of the dialog is to block and show that a cancellation is in progress. A better name might be `CANCELLATION_DIALOG`. `CANCEL_DIALOG` suggests to me that the user might be able to initiate some sort of cancellation.

**Review by dcbriccetti — CHANGES_REQUESTED** [review]

**Review by dcbriccetti — APPROVED** [review]
How about adding “caused by download cancellation” to the PR title?

This looks good. I did some additional related cleanup work, which you are free to incorporate into this PR, or leave for me for a new PR. https://github.com/dcbriccetti/collect/commit/1079b062ed19dfd789a806d4c85dc19d79502ee9

It would be good to get another reviewer here. (I’m not strong with Android.) @grzesiek2010, @shobhitagarwal1612, @jknightco? Who else?

**Comment by dcbriccetti** [comment]
Thanks for including my recent work in https://github.com/dcbriccetti/collect/commit/1079b062ed19dfd789a806d4c85dc19d79502ee9. About the commit comment, “Few changes in Exception handling”: Better might have been:

Include work from @dcbriccetti:

> Do some additional cleanup and simplification, motivated by current work to improve the reliability and user experience of canceling a form download.
> 
> No longer throw TaskCancelledException from findExistingOrCreateNewUri, since is it a very fast-running method. Improve its doc comment.
> Move the code that adds a new form to the database to a new method.

Or I think you could have cherry picked my commit to pull it into your PR. 

It looks like we’re getting to the end of this. Thanks for finding a solution. I still recommend we get more eyes on the change before it is merged.

**Comment by lakshyagupta21** [comment]
@dcbriccetti  Thanks for helping me with code improvements. I'll surely take care of these issues in future.

**Comment by yanokwa** [comment]
@srv-twry You had been able to reproduce this issue, so we'd love your review!

@lakshyagupta21 When you get a chance, please resolve the conflicts so it's a clean merge :)

File: collect_app/src/main/java/org/odk/collect/android/dao/FormsDao.java
Comment (dr0pdb) [inline collect_app/src/main/java/org/odk/collect/android/dao/FormsDao.java:149]: Do we need this ```finally``` block ?

**Review by dr0pdb — COMMENTED** [review]
Works fine locally and seems good to me :+1:  
Just a redundant block of code can be removed IMO.

File: collect_app/src/main/java/org/odk/collect/android/dao/FormsDao.java
Comment (lakshyagupta21) [reply collect_app/src/main/java/org/odk/collect/android/dao/FormsDao.java:149]: If any exception occurs then we need to close the cursor this is the reason `if (c != null)` is added just to make sure we close it.

Comment (dr0pdb) [reply collect_app/src/main/java/org/odk/collect/android/dao/FormsDao.java:149]: oh! okay

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java]: Putting a space after the comma here will address a checkstyle complaint.

**Review by dcbriccetti — CHANGES_REQUESTED** [review]

**Comment by dcbriccetti** [comment]
@yanokwa, do you think we have adequate review by Android experts?

**Review by dcbriccetti — APPROVED** [review]

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java]: This can continue being `final`.

File: collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java]: There's a `throw` after this so you don't need to set `fileResult`.

**Review by lognaturel — COMMENTED** [review]
Couple of small style comments inline but this is ready for blackbox testing. I think it has had enough eyes on it -- the approach is in line with other such operations in the code base.

Thanks @lakshyagupta21 and @dcbriccetti for the very thorough review!

**Comment by mmarciniak90** [comment]
@lognaturel @lakshyagupta21
I noticed that the user is moved back to main screen if the user cancels downloading a form.
Is it expected behavior? 
In the version without these changes, the user remains in the `Get Blank Form` view.

@opendatakit-bot label "needs discussion"

**Comment by grzesiek2010** [comment]
> I noticed that the user is moved back to main screen if the user cancels downloading a form.

I also don't like this behavior... is that the only possible solution?

**Comment by lakshyagupta21** [comment]
@mmarciniak90 @grzesiek2010 Now its resolved.

**Comment by dcbriccetti** [comment]
@mmarciniak90, once you’re done testing, I think we’re ready to merge.

**Comment by mmarciniak90** [comment]
**Tested with success!**
Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.0

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Comment by dcbriccetti** [comment]
Hurray! Great work, everybody.

**Pull request merged**
